Субачев Сергей

SergeySubachev opened the pull request:
Наставник - Карпова Оля @scklepova

SergeySubachev committed:
Class work

SergeySubachev committed:
Скопировал код решения задачи из блока Dependency Injection Container

SergeySubachev committed:
Переделал на Result<T>

scklepova commented on TagsCloud/FileParsers/DocParser.cs:
```diff
@@ -1,14 +1,14 @@
+        public Result<ImmutableList<string>> Parse(string filename)
         {
-            return new List<string> { "tag1", "tag1", "tag1", "tag2", "tag2", "tag3" };
+            return Result.Of(() => ImmutableList<string>.Empty.AddRange("Съешь ещё этих мягких французских булок да выпей же чаю".Split(' ')));
```
У `ImmutableList` есть более подходящий метод для создания коллекции. Но не конструктор, да. 
И замечание не только к этому месту в коде относится. Поищи ещё подобные использования.

SergeySubachev replied:
поправил

scklepova commented on TagsCloud/FileParsers/TxtParser.cs:
```diff
@@ -1,22 +1,21 @@
+
+            return Result.Of(() => File.ReadAllLines(filename))
+                .RefineError($"Can't read file '{filename}'.")
+                .Then(lines => ImmutableList<string>.Empty.AddRange(lines.SelectMany(line => line.Split(separators, StringSplitOptions.RemoveEmptyEntries))));
```
Почитай про метод `Split()` без аргументов. Это поможет тебе избавиться от списка разделителей (с ним очень легко пропустить какой-нибудь символ-разделитель. Например, у тебя нет двоеточия и точки с запятой)

SergeySubachev replied:
Почитал, удобно!, но не понял почему это работает...
В документации написано, что если аргументы пустые - разделяет по символам, которые IsWhiteSpace. Но в справке по IsWhiteSpace - ничего не говорится про знаки пунктуации... 
https://docs.microsoft.com/en-us/dotnet/api/system.char.iswhitespace?view=netframework-4.8
1) Почему они тоже учитываются?
2) Придется дополнительно проверять на пустые строки (если например два разделителя подряд будут), т.к. в StringSplitOptions это не получится указать.

scklepova replied:
1. Хм, про знаки препинания в этом методе я не задумывалась... вообще-то он не должен по ним делить, так что да, здесь не зайдет такое.
2. Эту проблему легко решить: можно просто передать пустой массив разделителей, и тогда можно использовать StringSplitOptions

SergeySubachev replied:
Нашел способ лучше - с помощью Regex

scklepova commented on TagsCloud/Layouters/CircularCloudLayouter.cs:
```diff
@@ -1,42 +1,48 @@
 
-                for (var angle = rnd.NextDouble() * Math.PI / 18; angle < 1.99 * Math.PI; angle += (Math.PI / 18))
+                var rnd = new Random();
+                for (var i = 1; i < sortedItems.Count; i++)
```
Можно добавить декларативности и заменить цикл на `Enumerable.Range`, как на паре учили

SergeySubachev replied:
Поправил, но вообще, на счет того, чтобы for менять Enumerable.Range - мне это не всегда нравится. А на паре, как ты говоришь, на которой это объясняли меня не было, я же не студент )) 
Есть какая-нибудь ссылка на презентацию, мотивирующую делать это? Хотелось бы посмотреть, потому что, видимо, это не всегда оправдано, по крайней мере, например, в учебных примерах, которые мы проходим на шпоре - вообще практически нигде нет такого...

scklepova replied:
Под парой я имела в виду занятие, на котором рассказывали про функциональное программирование. Презентация, по которой скорее всего вели занятие, лежит в том же репозитории, что и домашка.

SergeySubachev replied:
Ага, понял. Ну вот я и говорю, что кроме презентаций - такое в учебных проектах шпоры практически нигде увидишь больше)

SergeySubachev replied:
..хотя много таких мест типа:
for (int y = 0; y < Height; y++)
    for (int x = 0; x < Width; x++)

scklepova replied:
`Enumerable.Range` скорее просто альтернативный вариант написать `for`. Более fluent вариант. В продакш коде всё же чаще можно использовать `foreach`. С учебной целью - почему бы не попробовать Enumerable.Range)

scklepova commented on TagsCloud/Renderers/ColoredRenderer.cs:
```diff
@@ -20,46 +20,49 @@ private float FontSizeByTagRate(int minRate, int maxRate, int rate)
-        public override Image Render(List<LayoutItem> layoutItems)
+        public override Result<Image> Render(ImmutableList<LayoutItem> layoutItems)
         {
             if (layoutItems.Count() == 0)
```
Тут решарпер подсказывает использовать другой метод LINQ

SergeySubachev replied:
Решарпера нет, ну понял, да, это видимо чтобы всю коллекцию не перебирать - заменил на IsEmpty или !Any()

scklepova commented on TagsCloud/Renderers/ColoredRenderer.cs:
```diff
@@ -1,5 +1,5 @@
 ﻿using System;
```
Держи код в чистоте - убирай всё неиспользуемое, в том числе импорты. 
Кстати, у решарпера есть фича, позволяющая навести порядок в один хоткей. Погугли. Ключевые слова "code cleanup"

SergeySubachev replied:
С решарпером хорошо... Мне приходится постоянно тыкать контекстное меню "Remove and sort usings". Напрягает конечно... И, конечно, бывает забываю...

scklepova commented on TagsCloud/WordsFiltering/WasteWordsFilter.cs:
```diff
@@ -46,41 +47,43 @@ public class WasteWordsFilter : IFilter
+            return Task.Run(() => GetGrammems(words)).Result
+                .Then(grammems => ImmutableList<string>.Empty.AddRange(grammems
+                    .Select(grammem => grammem.Split(new char[] { '=', ',' }, 4))
+                    .Where(grInfo => (grInfo.Length == 2 || (grInfo.Length >= 3 && grInfo[2] != "сокр")) && allowDict.TryGetValue(grInfo[1], out var isAllow) && isAllow)
```
Допустимо писать несколько вызовов Where, чтобы не порождать длинные условия

SergeySubachev replied:
Поправил

scklepova commented on TagsCloud/WordsLoader.cs:
```diff
@@ -15,14 +14,13 @@ public WordsLoader(IFileParser[] parsers)
-                throw new ArgumentException($"Can't select file parser for this file format ({filename})");
-            return fileParser.Parse(filename);
+            return Result.Of(() => Path.GetExtension(filename))
+                .Then(fileExtension => parsers.FirstOrDefault(p => p.FileExtensions.Any(ext => ext == fileExtension)))
```
Если заменить вызов на First, можно будет не проверять на null парсер на следующем шаге. Если коллекция будет пустая, First бросит исключение, которое поймает .Then

SergeySubachev replied:
Поправил

scklepova commented on TagsCloud/WordsFiltering/WasteWordsFilter.cs:
```diff
@@ -46,41 +47,43 @@ public class WasteWordsFilter : IFilter
-            return new List<string>(File.ReadAllLines(output));
+                return output;
+            })
+                .Then(output => Result.Of(() => File.ReadAllLines(output)).RefineError($"Can't read file '{output}'."))
```
Если подобных кейсов много, можно создать перегрузку для .Then, которая будет принимать обычный Action или Func, а внутри оборачивать его вызов в Result.Of

SergeySubachev replied:
Нет, не надо. Then и так принимает обычный Action. Просто в данном случае я засунул Result.Of внутрь Then для того, чтобы конкретно у этого Result заменить сообщение об ошибке (RefineError), если не сможет прочитать файл output.

SergeySubachev committed:
Where(..&&..) разделил;
ImmutableList<>.Empty.AddRange заменил на Create и ToImmutableList;
for -> Enumerable.Range;
IEnumerable.Count() == 0 заменил на IsEmpty и !rectangles.Any();
FirstOrDefault -> First + ReplaceError;
Расширил Help у аргументов консольной версии программы - как вводить имена фильтров и др.;
SaveTo(Image) обернул в Result;
Метод Main декомпозировал и сделал одним Fluent-вызовом.

SergeySubachev committed:
Split - с учетом всех возможных разделителей и управляющих символов
